Add Managed Elasticsearch _bulk endpoint docs under managed inputs (#644)
elastic/opentelemetry: docs/reference/managed-inputs/authentication-delivery-and-failure-handling.md

---
navigation_title: Authentication, delivery, and failure handling
description: Learn how managed inputs share authentication, buffering and delivery, failure handling, and Elastic Cloud Hosted network limitations.
type: overview
applies_to:
  serverless: ga
  deployment:
    ech: ga
products:
  - id: cloud-serverless
  - id: cloud-hosted
  - id: observability
  - id: security
---

# Authentication, delivery, and failure handling with managed inputs [authentication-delivery-and-failure-handling]

> This page covers what all managed inputs share: how you authenticate, how data is briefly stored and delivered, what happens when indexing fails, and network limitations on {{ech}}. For protocol-specific setup, choose your endpoint in [Next steps](#next-steps).

Reviewer: ```suggestion
This page covers what all managed inputs share: how you authenticate, how data is stored for a limited time and delivered, what happens when indexing fails, and network limitations on {{ech}}. For protocol-specific setup, choose your endpoint in [Next steps](#next-steps).
```

It is at least 12h, so `briefly` is not very accurate

Author: ```suggestion
This page covers what all managed inputs share: how you authenticate, how data is stored for a limited time, how it is delivered, what happens when indexing fails, and network limitations on {{ech}}. For protocol-specific setup, choose your endpoint in [Next steps](#next-steps).
```

    ```json
    {
      "managed-inputs-writer": {
        "applications": [
          {
            "application": "apm",
            "resources": ["*"],
            "privileges": ["event:write"]
          }
        ]
      }
    }
    ```

4. Select **Create API key** and copy the encoded value.

For more details, refer to [Elasticsearch API keys](docs-content://deploy-manage/api-keys/elasticsearch-api-keys.md) and [Serverless project API keys](docs-content://deploy-manage/api-keys/serverless-project-api-keys.md).
:::

:::{dropdown} Using the Create API key API
Use the [Create API key API](https://www.elastic.co/docs/api/doc/elasticsearch-serverless/operation/operation-security-create-api-key). The same request works for both {{serverless-full}} and {{ech}}. For example, to create an API key named `managed-inputs-api-key`:

```console
POST /_security/api_key
{
  "name": "managed-inputs-api-key",
  "role_descriptors": {
    "managed-inputs-writer": {
      "applications": [
        {
          "application": "apm",
          "resources": ["*"],
          "privileges": ["event:write"]
        }
      ]
    }
  }
}
```
:::

> Send the API key in the `Authorization` header of each request to the managed endpoint as `ApiKey <api-key>`. For example:

Reviewer: The response is like this:

```
{
  "id": "TwWFxp8B7MvB7hK_GN8_",
  "name": "managed-inputs-api-key",
  "api_key": "_AO24gV5ToTNhcl8e3xFXg",
  "encoded": "VHdXRnhwOEI3TXZCN2hLX0dOOF86X0FPMjRnVjVUb1ROaGNsOGUzeEZYZw=="
}
```

```suggestion
Send the encoded API key in the `Authorization` header of each request to the managed endpoint as `ApiKey <encoded-api-key>`. For example:
```

```http
Authorization: ApiKey <api-key>
```

:::{note}
Index-level privilege scoping is not supported for managed inputs.
:::

## Buffering and delivery [delivery]

Managed inputs provide a durable ingest layer in front of {{es}}:

> - Incoming data is briefly stored (buffered) in a durable ingest layer before it reaches your {{es}} cluster. Buffered data is held for a limited time before it must be delivered.

Reviewer: ```suggestion
- Incoming data is stored (buffered) in a durable ingest layer before it reaches your {{es}} cluster. Buffered data is held for a limited time before it must be delivered.
```

Not briefly, at least 12h

> - When the service can't accept more data, endpoints apply back-pressure and respond with `429 Too Many Requests` so clients can retry. Refer to [Managed inputs rate limiting](rate-limiting.md).

Reviewer: ```suggestion
- When capacity controls reject data, endpoints can respond with `429 Too Many Requests`, so clients should retry with backoff. Other temporary service failures can return `503 Service Unavailable`. Refer to [Managed inputs rate limiting](rate-limiting.md).
```

---
navigation_title: Authentication, delivery, and failure handling
description: Learn how Elastic Cloud managed inputs share authentication, buffering and delivery, the failure store, and Elastic Cloud Hosted network limitations.
type: overview
applies_to:
  serverless: ga
  deployment:
    ech: ga
products:
  - id: cloud-serverless
  - id: cloud-hosted
  - id: observability
  - id: security
---

# Authentication, delivery, and failure handling with managed inputs [authentication-delivery-and-failure-handling]

This page covers what all managed inputs share: how you authenticate, how data is buffered and delivered, what happens when indexing fails, and network limitations on {{ech}}. For protocol-specific setup, choose your endpoint in [Next steps](#next-steps).

## Authentication [authentication]

> You can create an API key in the {{ecloud}} Console (for example from an **Add data** flow) or with the [{{es}} Create API key API](docs-content://deploy-manage/api-keys/elasticsearch-api-keys.md). The [Send data to the {{motlp}}](docs-content://solutions/observability/get-started/quickstart-elastic-cloud-otel-endpoint.md) quickstart walks through generating a pre-configured key. That same key works for all managed inputs.

Reviewer 2: Optional edit: I'd remove this since it interrupts the flow. Mid-sentence examples can useful deeper in the section/on the page, but tend to add extra text in overview paragraphs.
 
```suggestion
You can create an API key in the {{ecloud}} Console or with the [{{es}} Create API key API](docs-content://deploy-manage/api-keys/elasticsearch-api-keys.md). The [Send data to the {{motlp}}](docs-content://solutions/observability/get-started/quickstart-elastic-cloud-otel-endpoint.md) quickstart walks through generating a pre-configured key. That same key works for all managed inputs.
```

Reviewer 2: On a different note, I'm having trouble keeping track of the different types of API keys that users can create 😅 It seems like the first sentence is referring to the _Elasticsearch API keys_, which are different from the Elastic Cloud API keys for stack and serverless. If they are the ES API keys, it might be worth specifying that:

```suggestion
You can create an {{es}} API key in the {{ecloud}} Console (for example from an **Add data** flow) or with the [{{es}} Create API key API](docs-content://deploy-manage/api-keys/elasticsearch-api-keys.md). The [Send data to the {{motlp}}](docs-content://solutions/observability/get-started/quickstart-elastic-cloud-otel-endpoint.md) quickstart walks through generating a pre-configured key. That same key works for all managed inputs.
```

cc @shainaraskas in case you have more specific guidance on this.

Reviewer 2: As I'm reading through this PR, I'm also wondering if this is the managed inputs keys that's mentioned throughout, for example, [here](https://github.com/elastic/opentelemetry/pull/644/changes#diff-6cf5cef03d17d84781ee0b2d8c2e3effc70e2d510b6ff13b6a3d58873b1551cfR94).

Reviewer 3: yeah this definitely matters ... it is tied to [the key you create](https://www.elastic.co/docs/deploy-manage/api-keys) - this docset is the source of truth for terminology

this seems to be referring to two key types, because you can't create an ES API key directly from the ecloud console ... you can create a cloud API key with elasticsearch access

if these managed inputs are per-project/deployment, prob the following key types would work

* elasticsearch key created within the serverless project or ECH deployment. THESE ARE DOCUMENTED SEPARATELY FOR DUMB REASONS
	* serverless: https://elastic.co/docs/deploy-manage/api-keys/serverless-project-api-keys
	* elasticsearch (includes ECH): https://elastic.co/docs/deploy-manage/api-keys/elasticsearch-api-keys
* elastic cloud key with ES/kib access (SERVERLESS ONLY)
	* https://elastic.co/docs/deploy-manage/api-keys/elastic-cloud-api-keys#project-access
	
as a side note, we shouldn't say "Elastic Cloud" just naked, because that's just a host for projects/deployments. instead:

> Managed inputs are fully managed ingestion endpoints for sending data to projects and deployments hosted on Elastic Cloud.

Author: Thanks both! I reworked the Authentication section in  https://github.com/elastic/opentelemetry/pull/644/commits/cd12d402245791b93de6d5468b8f0030841e9623.

> To send data to any managed input, your API key must include the `event:write` privilege for the `apm` application. You use the same privilege for every managed input endpoint, including OpenTelemetry Protocol (OTLP), Prometheus Remote Write, and `_bulk`.

Reviewer 2: I'm not familiar with this feature at all, so I can't tell whether the second sentence should be a statement or guidance (e.g., "You should use the same privilege for every managed input endpoint...").

Author: Reworked this section in https://github.com/elastic/opentelemetry/pull/644/commits/cd12d402245791b93de6d5468b8f0030841e9623

The following example creates a `managed_inputs_writer` role with the required privilege:

```json
{
  "managed_inputs_writer": {
    "applications": [
      {
        "application": "apm",

>         "resources": ["*"],

Reviewer 4: Hello @perk 

The `resource` set to `*` is associated to the name of the final data streams where the data will be stored on?

How can the Managed ES Bulk endpoint authenticate all the requests? A bulk request can contain multiple data streams, aliases or inputs. How we authenticate ll the requests?

What is the permission model?

Reviewer: Data streams don't matter at this point.

We just expect that the customer project/deployment has an API key with these permissions. 

We don't support granular API keys in any form (per data stream, per source, etc)

        "privileges": ["event:write"]
      }
    ]
  }
}
```

> Send the key in each request's `Authorization` header as `ApiKey <api-key>`.

Reviewer 2: An example could be useful here and would illustrate the exact format of the request. I'm not sure if [these](https://www.elastic.co/docs/api/doc/elasticsearch/authentication#doc-authentication-api-key-auth) are the corresponding API docs, but sharing in case they are.

Same comment applies [here](https://github.com/elastic/opentelemetry/pull/644/changes#diff-6cf5cef03d17d84781ee0b2d8c2e3effc70e2d510b6ff13b6a3d58873b1551cfR79).

Author: Addressed in https://github.com/elastic/opentelemetry/pull/644/commits/cd12d402245791b93de6d5468b8f0030841e9623 
I think it depends on the shipper and its configuration, so keeping this at the minimum…

:::{note}

> Index-level privilege scoping is not supported for managed inputs. API keys restricted to specific index-level privileges return a `PermissionDenied` error.

Reviewer 2: Totally optional edit. I think the first sentence is strong enough to stand on it's own. 

Plus, the second sentence almost suggests that you _can_ set index-level privs, but you can't actually make requests against those indices. Removing the second sentence eliminates any possibility that you can/should do either.

```suggestion
Index-level privilege scoping is not supported for managed inputs.
```

:::

## Buffering and delivery [delivery]

Managed inputs provide a durable ingest layer in front of {{es}}:

> - Incoming data is buffered before it reaches your {{es}} cluster. Buffered data is held for a limited time before it must be delivered.

Reviewer 2: Buffered is a new term to me! What is it and has the term been defined in the docs elsewhere?

Author: (Hopefully) addressed in https://github.com/elastic/opentelemetry/pull/644/commits/cd12d402245791b93de6d5468b8f0030841e9623

- When the service can't accept more data, endpoints apply back-pressure and respond with `429 Too Many Requests` so clients can retry. Refer to [Managed inputs rate limiting](rate-limiting.md).

> - A successful accept response means managed inputs durably accepted the data for processing, not that {{es}} has finished indexing it.

Reviewer 2: This is identical to the first sentence in the next section, but I don't really understand how they're connected. Can it be removed or reframed?

```suggestion
```

:::{note}

> For the Managed {{es}} _bulk endpoint, a batch is atomic: the endpoint accepts or rejects the whole batch, and a `201` per item means the data is durably enqueued, not indexed. For details, refer to [Delivery behavior](elasticsearch-bulk.md#delivery-behavior).

Reviewer 2: ```suggestion
For the Managed {{es}} `_bulk` endpoint, a batch is atomic: the endpoint accepts or rejects the whole batch, and a `201` per item means the data is durably enqueued, not indexed. For details, refer to [Delivery behavior](elasticsearch-bulk.md#delivery-behavior).
```

Reviewer 2: Resolving - @vishaangelova is using the same styling as the UI (refer to https://github.com/elastic/cloud-ui/pull/3764).

:::

## Indexing errors and the failure store [failure-store]

A successful accept response from a managed input means the data was durably accepted for processing, not that {{es}} has indexed it. Indexing errors, such as mapping conflicts or ingest pipeline errors, can happen asynchronously after the data is accepted, and aren't reported back to the client.

Managed inputs don't enable or manage the [failure store](docs-content://manage-data/data-store/data-streams/failure-store.md). The failure store is an {{es}} data stream setting. If the destination data stream has it enabled, documents that fail indexing are written there. If it isn't enabled, those documents aren't captured.

:::{warning}
If a document fails indexing and the destination data stream doesn't have the failure store enabled, the document is dropped. Because indexing errors happen after the data is accepted, your shipper still reports success, so this data loss is silent.
:::

> To confirm your data was indexed, verify that documents landed in the destination data stream, and use [Data Set Quality](docs-content://solutions/observability/data-set-quality-monitoring.md) to monitor and triage indexing issues.

Reviewer 2: The guidance on ensuring your data was index seems important enough to place at the start of the section. Right now, the first para mainly tells the reader what the accept response _isn't_.

Author: Agreed. Addressed in  https://github.com/elastic/opentelemetry/pull/644/commits/cd12d402245791b93de6d5468b8f0030841e9623

## {{ech}} limitations [ech-limitations]

```{applies_to}
ech:
```

In {{ech}} deployments, the following limitations apply to managed inputs:

- [IP filters](docs-content://deploy-manage/security/ip-filtering-cloud.md) do not apply to managed endpoints.
- Managed endpoints are not available over a [private connection](docs-content://deploy-manage/security/private-connectivity.md). When private connectivity is configured, the public managed endpoint is still available.

## Next steps [next-steps]

After you understand authentication and delivery, configure the endpoint for your protocol:

- [Ingest OpenTelemetry data with the Managed OTLP endpoint](managed-otlp-endpoint.md)
- [Ingest Prometheus metrics with the Managed Prometheus Remote Write endpoint](prometheus-remote-write.md)

> - [Ingest data with the Managed {{es}} _bulk endpoint](elasticsearch-bulk.md)

Reviewer 2: ```suggestion
- [Ingest data with the Managed {{es}} `_bulk` endpoint](elasticsearch-bulk.md)
```

Reviewer 2: Resolving - uses the same styling as the UI.

:::

## Buffering and delivery [delivery]

Managed inputs provide a durable ingest layer in front of {{es}}:

- Incoming data is buffered before it reaches your {{es}} cluster.
- When the service can't accept more data, endpoints apply back-pressure and respond with `429 Too Many Requests` so clients can retry. Refer to [Managed inputs rate limiting](rate-limiting.md).
- A successful accept response means managed inputs durably accepted the data for processing, not that {{es}} has finished indexing it.

:::{note}
For the Managed {{es}} _bulk endpoint, a batch is atomic: the endpoint accepts or rejects the whole batch, and a `201` per item means the data is durably enqueued, not indexed. For details, refer to [Delivery behavior](elasticsearch-bulk.md#delivery-behavior).
:::

## Failure store [failure-store]

```{applies_to}
stack: ga 9.1+
```

Managed inputs are built for high availability, but some failures can still prevent events from being written to your {{es}} cluster (for example, ingest pipeline errors or mapping conflicts).

> To protect data in these cases, managed inputs use the [Failure store](docs-content://manage-data/data-store/data-streams/failure-store.md). For managed inputs data streams, the failure store is always enabled.

Reviewer 4: The PM can keep me honest here... 

If I take an Elastic Agent, I deploy the system integration on it and I point it to the Managed ES Bulk API, the Fleet index templates of `metrics-system*` data streams do not have Failure Store enabled.

Therefore I struggle to understand how just the fact of using Managed ES Bulk endpoint would make this work...

@nimarezainia / @perk can you clarify?

Author: @constanca-m, do the failure store and the `stack: ga 9.1+` applicability apply identically to Managed Prometheus Remote Write and Managed ES _bulk data streams?

I had inferred the failure store applicability for these endpoints, but hadn’t confirmed it, so this is probably wrong... Please correct me here.

Reviewer: Failure store is not a Hosted OTel Collector feature, but an Elasticsearch data stream setting, right? I don't see why this is mixed with the managed OTLP endpoints.

I wouldn't write anything about it, personally. On managed OTLP, we don't do anything to handle the failure store automatically, nor is it part of the plan, as far as I am aware

Reviewer 4: Yes exactly my point here to clarify is: we mention shipping data to Managed ES Bulk endpoint will "automatically" make use of the Failure Store on ES side.

Afaik this is not generally true.

The use of failure store on the destination data streams is driven by several conditions (mostly dependent on the defaults and customizations set on the destination cluster)
- if the destination is ECH deployment or Serverless project (as we enable 
- if the user set the cluster setting https://www.elastic.co/docs/manage-data/data-store/data-streams/failure-store#set-up-failure-store-cluster-setting
- if the user modified the data stream or index template of the data stream
- if the default index template of the integration enabled it

Just to expand with one example.
**Situation: Elastic Agent managed by Fleet with the System integration on ECH.**
The destination data stream for the System integration are mostly `metrics-system*` data streams and `logs-system*` data streams.

Afaik, this is our status now:
```
Which data streams are affected by default (9.2.0+)

┌─────────────────────┬───────────────────────┬──────────────────────────────────┐
│ Data stream pattern │ Failure store default │       Template responsible       │
├─────────────────────┼───────────────────────┼──────────────────────────────────┤
│ logs-*              │ Enabled (from 9.2.0)  │ logs@settings                    │
├─────────────────────┼───────────────────────┼──────────────────────────────────┤
│ metrics-*           │ Disabled              │ metrics@settings (no FS setting) │
├─────────────────────┼───────────────────────┼──────────────────────────────────┤
│ traces-*            │ Disabled              │ traces@settings (no FS setting)  │
├─────────────────────┼───────────────────────┼──────────────────────────────────┤
│ Custom data streams │ Disabled              │ Must be configured manually      │
└─────────────────────┴───────────────────────┴──────────────────────────────────┘
```

Note: `otel` and `prometheus` data streams have FS enabled too since 9.2

So to me the fact of using the Failure Store is totally dependant on the destination cluster.

Did we evaluate the fact using Managed Bulk ES endpoint can possibly lead to data loss for the `metrics-*` (if they're not `otel` or `prometheus`?

Reviewer: > Did we evaluate the fact using Managed Bulk ES endpoint can possibly lead to data loss for the metrics-* (if they're not otel or prometheus?

Bulk endpoint is only enabled for logs

Author: (Hopefully) addressed in https://github.com/elastic/opentelemetry/commit/45d883480764a044dedb295bb281877974dc80db and https://github.com/elastic/opentelemetry/commit/123f3ad2b722e7b20c552a5409230e65f0dda801.
